python/cudf/cudf/core/cut.py

```diff
@@ -0,0 +1,277 @@
+    ...       dtype='category')
+    >>> cudf.cut(np.array([1, 7, 5, 4, 6, 3]), 3,
+    ...       labels=["B", "A", "B"], ordered=False)
+    CategoricalIndex(['B', 'B', 'A', 'A', 'B', 'B'], categories=['A', 'B'],
+    ...        ordered=False, dtype='category')
+    >>> cudf.cut([0, 1, 1, 2], bins=4, labels=False)
+    array([0, 1, 1, 3], dtype=int32)
+    Passing a Series as an input returns a Series with categorical dtype:
+    >>> s = pd.Series(np.array([2, 4, 6, 8, 10]),
+    ...        index=['a', 'b', 'c', 'd', 'e'])
+    >>> pd.cut(s, 3)
+    """
+    left_inclusive = False
+    right_inclusive = True
+    # saving the original input x for use in case its a series
+    orig_x = x
+
+    if not ordered and labels is None:
+        raise ValueError("'labels' must be provided if 'ordered = False'")
+
+    if duplicates not in ["raise", "drop"]:
+        raise ValueError(
+            "invalid value for 'duplicates' parameter, valid options are: "
+            "raise, drop"
+        )
+
+    # bins can either be an int, sequence of scalars or an intervalIndex
+    if isinstance(bins, Sequence):
+        if len(set(bins)) is not len(bins):
+            if duplicates == "raise":
+                raise ValueError(
+                    f"Bin edges must be unique: {repr(bins)}.\n"
+                    f"You can drop duplicate edges by setting the 'duplicates'"
+                    "kwarg"
+                )
+            elif duplicates == "drop":
+                # get unique values but maintain list dtype
+                bins = cupy.unique(bins).tolist()
+
+    # if bins is an intervalIndex we ignore the value of right
+    if (
+        right is False
+        and isinstance(bins, pandas_IntervalIndex)
+        and bins.closed == "right"
+    ):
+        right = True
+
+    # create bins if given an int or single scalar
+    if not isinstance(bins, pandas_IntervalIndex):
+        if not isinstance(bins, (Sequence)):
+            if isinstance(
+                x, (pd.Series, cudf.Series, np.ndarray, cupy.ndarray)
+            ):
+                # changing to scalars and using sequence to get the bins
+                # because this allows masked arrays from value_counts to
+                # also be able to be handled by cut correctly
+                # pandas by default seems to turn all bins into a float
+                mn = cudf.Scalar(x.min(), dtype="float64")
+                mx = cudf.Scalar(x.max(), dtype="float64")
+            else:
+                mn = cudf.Scalar(min(x), dtype="float64")
+                mx = cudf.Scalar(max(x), dtype="float64")
+            step = cudf.Scalar((mx - mn) / bins, dtype="float64")
+            bins = sequence(
+                size=bins + 1, init=mn.device_value, step=step.device_value
+            ).values
+            adj = (mx - mn).value * 0.001
+            if right:
+                bins[0] -= adj
+            else:
+                bins[-1] += adj
```

Comment:
Looks like we don't really need to perform the sequence making on device... can we just compute the whole thing on host? Looking down the lines there are other operations on `bins` as well.

Also, is the bin edge handling correct? `if right` means the rightmost edge is inclusive, shouldn't we adjust that instead?

Reply:
We probably could perform everything on host, I just ran into some errors when I was working with `series.value_counts`, when I was trying to use `cudf.cut` to get the intervals given bins. The main issue was working with masked arrays. I'll check if I can change this to work on the host! 

In terms of the bin edge handling, I'm imitating pandas behavior, which adjusts both the first bin element and the rightmost edge for some reason 😄

Reply:
I'm just seeing some of these comments today @isVoid not sure how I missed them lol. Anyway, I've made a few more updates and will try and address these. If you spot anything else, let me know! Hoping to merge this next week! Also please flag where you see some device syncing happening that I might have missed. Thank you :D